### ISSUE TITLE
Generation of portable models with file dependencies on external resources

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/ModelicaSpecification#1676]
**Reported by beutlich on 10 Mar 2015 11:41 UTC**
Imagine the code export as FMU of a model that has external file dependencies, like e.g. 
```mo
model MyFMU
  Modelica.Blocks.Sources.CombiTimeTable combiTimeTable(
    tableOnFile=true,
    tableName="a",
    fileName="C:/MyProjects/Libraries/Modelica 3.2.1/Resources/Data/Tables/test.txt",
    columns=2:2)
    annotation (Placement(transformation(extent={{-56,30},{-36,50}})));
end MyFMU;
```
Whereas the transient simulation of such a model runs without problems - given that combiTimeTable.fileName does exist and is a valid file - the exported FMU is no longer portable (to a different platform or computer) because test.txt cannot be found.

Our idea is to have a new code-export annotation (could be also a custom annotation) on Modelica.Blocks.Sources.CombiTimeTable.fileName that this parameter is a file dependency for the translation of the model. The translation mode (e.g. transient, FMU or special target) then decides how to use this new annotation. In case of an FMU it would be wise to add text.txt to the resources directory of the FMU such that an FMU siomulator can consider it at import or simulation.

Please not that this is different from the loadResource function. The goal of this issue is to automatically generate a<ins>portable</ins> translated model (e.g. FMU) that runs out-of-the-box and gives the same simulation results as the non-portable transient simulation.

What do you think? Is it useful to add such an code-export annotation to the MLS or should it be left to the tool vendors by considering either a vendor-specific or custom annotation? See MLS 3.3 rev 1, section 18.3 for standardized annotations for code generation.

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/1676

**Comment by modelica-trac-importer** [comment]
**Comment by choeger on 10 Mar 2015 12:10 UTC**
Personally, I do not like the idea very much. Stuff like this is the traditional job of buildsystems (automake,cmake,vs etc). The issue you propose to fix is already non-portable at the source level and should be avoided.

However, I do understand that there is currently no such thing as a platform independent modelica build system. IMO the experiment annotation is what comes closest to it today. So how about moving that parameter (the absolute directory e.g. "C:/MyProjects/Libraries/Modelica 3.2.1/Resources/") to the experiment annotation and specify your proposed handling there?

**Comment by modelica-trac-importer** [comment]
**Comment by jmattsson on 10 Mar 2015 12:12 UTC**
I'd say that the loadResource function _should_ have been defined as a built-in function (what did we decide to call them again?) with this behavior. Interpreting modelica URIs is dependent of the translation process anyway, so the loadResource function has to be implemented as a built-in function anyway. Also, It wouldn't be a problem to let it have a dual purpose, since it would be pretty useless to use a modelica URI to a file in a model compiled to an FMU without including the file.

My suggestion would be to introduce a built-in function with this behavior, and deprecated the loadResource function in favor of the new built-in function.
The behavior would be:
 * takes an absolute path, a modelica URI or a file URI
 * adds the file to the resource folder when translating to an FMU (or whatever else is needed to make sure it is available for other targets)
 * returns an absolute path to the file (oringinal file, in resource folder, etc - depending on translation target) during simulation

**Comment by modelica-trac-importer** [comment]
**Comment by oelsner on 10 Mar 2015 12:41 UTC**
> Stuff like this is the traditional job of buildsystems

I agree, but buildsystems must get a hint, that there is a reference, 
which requires external files. 
This s similar to the library/include annotation (12.9.4) but for parameters.

> My suggestion would be to introduce a built-in function with this  behavior.

The translation process can decide, how to handle this information.
I think a built-in function is not absolutely necessary.

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 10 Mar 2015 21:21 UTC**
> IMO the experiment annotation is what comes closest to it today. So how about moving that parameter (the absolute directory e.g. "C:/MyProjects/Libraries/Modelica 3.2.1/Resources/") to the experiment annotation and specify your proposed handling there?

Someone has to collect all the parameter resource information and fill the model experiment annotation. This is neither object-oriented nor handy. Having it already in the library types (instead of in the model under development) is much more user-friendly.

**Comment by modelica-trac-importer** [comment]
**Comment by jmattsson on 11 Mar 2015 09:00 UTC**
Replying to [comment:3 oelsner@…]:
> > My suggestion would be to introduce a built-in function with this  behavior.
> The translation process can decide, how to handle this information.
> I think a built-in function is not absolutely necessary.

Not necessary, but it makes more sense and is clearer to the user than having an annotation that means that the tool in some situations should change the value of the parameter. Also, it's more flexible, since you could then use it directly in e.g. the constructor of an external object without having a parameter at all.

**Comment by modelica-trac-importer** [comment]
**Comment by choeger on 11 Mar 2015 09:22 UTC**
Replying to [comment:4 beutlich]:
> This is neither object-oriented nor handy. Having it already in the library types (instead of in the model under development) is much more user-friendly.

Absolute paths in models is neither portable nor handy ;). In fact, a model with an absolute path in it must either only use that absolute path at the toplevel (i.e. where the experiment annotation is and another user is capable to find it) or will only work on at the site-of-creation. I do not see the point in having "portable" FMUs at the price of non-portable Modelica models.

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 11 Mar 2015 09:25 UTC**
> The issue you propose to fix is already non-portable at the source level and should be avoided.

OK, should be avoided. But let's consider this as given and assume we have the non-portable model and cannot change this.

**Comment by modelica-trac-importer** [comment]
**Comment by choeger on 11 Mar 2015 10:04 UTC**
Replying to [comment:7 beutlich]:
> > The issue you propose to fix is already non-portable at the source level and should be avoided.
> OK, should be avoided. But let's consider this as given and assume we have the non-portable model and cannot change this.

Then we should not support it any further. In your proposal, the user of such models would have to execute an action (add that annotation). However, to do so, he has to accept that there is a problem in the first place. When that conclusion is reached, the proper course of action is to use a relative path (solving the underlying issue), not to try to patch the problem.

**Comment by modelica-trac-importer** [comment]
**Comment by c&#x2e;schulze on 11 Mar 2015 10:57 UTC**
Just to add some of our problems with external data files... Our customers often have problems with missing licenses for external libraries, missing data files, and missing dlls.

* We have data files, that are loaded based on an absolute filename (some combi-table like implementation, but it is an external dll called with external functions)

* We have data files for medium properties, which are loaded based on a medium name

* We have external libraries that internally load a dll, which require a specific version of additional dlls to be loaded prior to them.

There is currently no specified way to tell the Modelica-Tool which additional external files have to be included in an FMU. We have to provide an additional postprocessing tool.

To make the models work we currently search these external data files and dlls
* relatively to the loaded dll (to check whether it is running in some FMU and have locally stored files in Resources...), 
* relatively to the executing program (in case the user has copyied the dll to the system path and uses some dymosim.exe like environment)
* in a predefined %allusers%-path
* in a path set by an environment variable

... that is a pain.
**Whatever you're planning to do, it would be nice to pass some information to the external libraries/dlls, where the resources folder is placed.**
For external data files an URI-Resources-Path-solution that could be accessed by external libraries at runtime would be awesome. And it would be nice to be able to tell the Modelica-Tool which files to are to be included in an FMU or SaveTotal.

**Comment by modelica-trac-importer** [comment]
**Comment by jmattsson on 12 Mar 2015 11:30 UTC**
Replying to [comment:1 choeger]:
> The issue you propose to fix is already non-portable at the source level and should be avoided.

Many models need external data files, that is a fact of life. Shouldn't we then strive to support such models in a way that is as portable as possible? I would like it to be possible to write models that use data files and are source-portable. I would also like it to be possible and preferably convenient to write a model that use data files and that you do not intend to share the source of, compile it into an FMU and send it to someone else.

**Comment by modelica-trac-importer** [comment]
**Comment by choeger on 12 Mar 2015 12:10 UTC**
Replying to [comment:10 jmattsson]:
> Replying to [comment:1 choeger]:
> > The issue you propose to fix is already non-portable at the source level and should be avoided.
> Many models need external data files, that is a fact of life. Shouldn't we then strive to support such models in a way that is as portable as possible? I would like it to be possible to write models that use data files and are source-portable. I would also like it to be possible and preferably convenient to write a model that use data files and that you do not intend to share the source of, compile it into an FMU and send it to someone else.

Yes, of course. That reasoning is perfectly sound. However, the external files should not be referenced via some absolute path but a relative one. If the resolution works via a standardized lookup mechanism, a (builtin) function or some other way I do not care. But I oppose the idea of having absolute paths in source files because that will never be a portable solution (and I find the idea that the FMU is more portable then its source somewhat disturbing).

**Comment by modelica-trac-importer** [comment]
**Comment by jmattsson on 12 Mar 2015 13:10 UTC**
Replying to [comment:11 choeger]:
> However, the external files should not be referenced via some absolute path but a relative one. [...] But I oppose the idea of having absolute paths in source files because that will never be a portable solution

I'm not saying that absolute paths should be the primary option, but it would be good to have it as _a_ option. Relative paths (that are not modelica-URIs) are problematical in Modelica code since it is a bit unclear what they are relative _to_.

Replying to [comment:11 choeger]:
> and I find the idea that the FMU is more portable then its source somewhat disturbing.

Disturbing, but useful. :)

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 12 Mar 2015 14:56 UTC**
Replying to [ticket:1676 beutlich]:
> Whereas the transient simulation of such a model runs without problems - given that combiTimeTable.fileName does exist and is a valid file - the exported FMU is no longer portable (to a different platform or computer) because test.txt cannot be found.
> 
> Our idea is to have a new code-export annotation (could be also a custom annotation) on Modelica.Blocks.Sources.CombiTimeTable.fileName that this parameter is a file dependency for the translation of the model. The translation mode (e.g. transient, FMU or special target) then decides how to use this new annotation. In case of an FMU it would be wise to add text.txt to the resources directory of the FMU such that an FMU siomulator can consider it at import or simulation.
> 
> Please not that this is different from the loadResource function. The goal of this issue is to automatically generate a<ins>portable</ins> translated model (e.g. FMU) that runs out-of-the-box and gives the same simulation results as the non-portable transient simulation.
> 
> What do you think? Is it useful to add such an code-export annotation to the MLS or should it be left to the tool vendors by considering either a vendor-specific or custom annotation? See MLS 3.3 rev 1, section 18.3 for standardized annotations for code generation.

  
You are right that this problem should be solved, so that there is a hint in the library what a tool is expected to do and then tool specific support for it.

I wonder why loadResource should not be used for it? loadResource requires as input an URI, such as "modelica://MyLibrary/Resources/Maps/table1.txt" and returns the absolute file name for it. If the resource is stored in a data base system, this might require to extract the data, store it in the file system and return the absolute file name.

One could include the calling of loadResource in the table blocks (or in generally, whenever a file name is required in MSL, then the file name is first processed with loadResource).

When an FMU is generated, a tool is processing the Modelica model and when it detects a loadResource call, it evaluates the input argument during translation, stores the file in the Resources directory of the FMU, and returns the absolute file name in the Resources directory in the environment where the FMU is imported. This is possible, because during instantiation of the FMU, the Resources directory is passed as input argument to the FMU and since the FMU knows where the data file is relative to Resources directory, it can construct the full path name and return this name from the loadResource function call.

To summarize: From Modelica Association point of view, one should identify the places in the MSL where a file is read, and use loadResource at this place. It is then up to the tool vendors, whether they support the above approach.

Introducing a new annotation or a built-in function looks not so attractive to me, because we have the mechanism already and then suddenly both the tools and the users have to understand two similar (but probably not identical) mechanisms for file handling and this can become quite confusing.

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 12 Mar 2015 15:11 UTC**
Thank you all for your contributions. Please do not wonder why ITI is quiet today. We have prepared a list of use cases and proposals to be discussed at an inhouse meeting scheduled for March, 16th with our Modelica/FMU experts. And yes, loadResource is used there.

**Comment by modelica-trac-importer** [comment]
**Comment by jmattsson on 12 Mar 2015 15:21 UTC**
Replying to [comment:13 otter]:
> I wonder why loadResource should not be used for it? 

For the modelica-URI case loadResource could be used - this is in fact the current behavior for JModelica.org when a modelica-URI is used as argument to loadResource (including the file in the FMU, resolving the URI to a place in the Resources folder). However, if we want to allow also regular paths or file-URIs, then it can't be used since it defines a different behavior for those cases. Also, it is pretty inconsistent to have a function in ModelicaServices that influence the behavior of the compiler that much.

**Comment by modelica-trac-importer** [comment]
**Comment by oelsner on 17 Mar 2015 07:28 UTC**
Our internal dicussion indicated that we will handle this problem as a tool issue.
We rejected the idea of a code generation annotation, because it's not clear whether a reference should be included or not. Data in external files could be confidential.

The user has to define, which files has to be included in the portable model.
When the user decides to include a file, then full paths will be replaced by an URI 
which defines the location of the file relative to the container.
These decisions are stored at experiment level.

__Example for FMU Export:__

Modification in modelica ...

```
file="C:\temp\data.csv"
```
 
will be replaced in modelDescription.xml of the FMU by ...

```
file="fmu://resources/data.csv"
```

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 1 Apr 2016 12:25 UTC**
Replying to [comment:16 oelsner@…]:
> Our internal dicussion indicated that we will handle this problem as a tool issue.
> We rejected the idea of a code generation annotation, because it's not clear whether a reference should be included or not. Data in external files could be confidential.

Based on this I propose to close the ticket as worksforme.
(and I know that other tools also can handle variations of this).

**Comment by modelica-trac-importer** [comment]
**Modified by beutlich on 1 Apr 2016 12:27 UTC**
